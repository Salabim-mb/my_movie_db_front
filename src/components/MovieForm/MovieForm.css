.react-datepicker-wrapper {
    display: block;
}

.spinner {
    justify-content: center;
    display: flex;
}