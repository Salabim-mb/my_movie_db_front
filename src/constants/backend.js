const backend_path = process.env.REACT_APP_BACKEND_PATH;

export const backend = {
    MOVIES: backend_path + "movies/",
    USERS: backend_path + "user/",
    PLAIN: backend_path,

};