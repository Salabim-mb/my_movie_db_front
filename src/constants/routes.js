import MovieList from "../components/MovieList";

const paths = {
    MAIN: "/",
    LOGIN: "/login",
    REGISTER: "/register",
    MOVIES: "/movies",
    NEW_MOVIE: "/movies/new"
};

export default [
    {
        path: paths.MOVIES,
        exact: true,
        component: MovieList
    }r
]

export {paths};