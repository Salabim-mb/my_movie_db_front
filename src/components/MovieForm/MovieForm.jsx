import React, {useEffect, useState} from "react";
import {Alert, Button, Form, Spinner} from "react-bootstrap";
import {useParams} from "react-router-dom";
import Page from "../Page";
import {fileToBase64} from "../../utils/FileToBase64";
import DatePicker from 'react-datepicker';
import "react-datepicker/dist/react-datepicker.css";
import "./MovieForm.css";
import {backend} from "../../constants/backend";
import {getHeaders} from "../../utils/CORSHeaders";

const sendMovie = async (method, data, id) => {
    let baseUrl = backend.MOVIES;
    if (method === "PUT") {
        baseUrl += `${id}/`;
    }
    const headers = getHeaders();
    let res = await fetch(baseUrl, {
        method: method,
        headers,
        body: JSON.stringify(data)
    });

    if (res.status === 200) {
        return await res.json();
    } else {
        throw res.status;
    }
};

const getMovieDetails = async (id) => {
    let baseUrl = backend.MOVIES + id + "/";
    const headers = getHeaders();

    let res = await fetch(baseUrl, {
        method: "GET",
        headers
    });

    if (res.status === 200) {
        return await res.json();
    } else {
        throw res.status;
    }
};

const mapResponseData = (source) => ({
    name: source.movieName,
    genre: source.movieGenre,
    director: source.movieDirector,
    description: source.movieDescription,
    releaseYear: new Date( source.movieReleaseYear ),
    image: source?.movieImage || null
});

const MovieForm = () => {
    const [data, setData] = useState({
        name: "",
        director: "",
        genre: "",
        releaseYear: undefined,
        image: null,
        description: ""
    });
    const [photoLabel, setPhotoLabel] = useState(null);
    const [method, setMethod] = useState("POST");
    const [loading, setLoading] = useState(false);
    const [photoError, setPhotoError] = useState(false);
    const [error, setError] = useState(false);
    const [getError, setGetError] = useState(false);
    const [validated, setValidated] = useState(false);
    const [resCorrect, setResCorrect] = useState(false);
    const [disabled, setDisabled] = useState(false);

    let fileInput = React.createRef();

    const location = useParams();

    useEffect(() => {
        const loadMovie = async () => {
            setLoading(true);
            try {
                let res = await getMovieDetails(location.id);
                setData(mapResponseData(res));
                setMethod("PUT");
            } catch(e) {
                console.log(e);
                setGetError(true);
            } finally {
                setLoading(false);
            }
        };
        if (location?.id) {
            loadMovie();
        }
    }, [location.id]);

    const prepareData = () => ({
        movieTitle: data.title,
        movieDirector: data.director,
        movieGenre: data.genre,
        movieReleaseYear: data.releaseYear.toISOString(),
        movieDescription: data.description,
        movieImage: data.image
    });

    const handleSubmit = async (event) => {
        setError(false);
        setDisabled(true);
        event.preventDefault();
        const form = event.currentTarget;
        if (form.checkValidity() === false) {
            event.stopPropagation();
        } else {
            setValidated(true);
            try {
                prepareData();
                await sendMovie(method, data, location?.id);
                setResCorrect(true);
            } catch(e) {
                console.log(e);
                setError(true);
            }
        }
        setDisabled(false);
    };

    const onPhotoChange = async () => {
        setPhotoError(false);
        const file = fileInput.current?.files?.[0];
        try {
            let b64 = await fileToBase64(file);
            setData({...data, image: b64});
            setPhotoLabel(file.name);
        } catch(e) {
            setPhotoError(true);
        }
        fileInput = null;
    };

    return (
        <Page>
            {
                loading ? (
                <div className="spinner">
                    <Spinner animation="border" role="status" />
                </div>
                ) : getError ? (
                    <Alert variant="danger">Something went wrong while trying to get movie data.</Alert>
                ) : (
                    <Form noValidate validated={validated} onSubmit={handleSubmit}>
                        <Form.Group>
                            <Form.Label>Movie title</Form.Label>
                            <Form.Control
                                value={data.name}
                                onChange={(e) => setData({...data, name: e.target.value})}
                                name="name"
                                required
                                placeholder="Interstellar"
                                type="text"
                            />
                            <Form.Control.Feedback type="invalid">
                                Please provide movie title
                            </Form.Control.Feedback>
                        </Form.Group>
                        <Form.Group>
                            <Form.Label>Director</Form.Label>
                            <Form.Control
                                value={data.director}
                                onChange={(e) => setData({...data, director: e.target.value})}
                                name="director"
                                required
                                placeholder="Christopher Nolan"
                                type="text"
                            />
                            <Form.Control.Feedback type="invalid">
                                Please provide movie director
                            </Form.Control.Feedback>
                        </Form.Group>
                        <Form.Group>
                            <Form.Label>Genre</Form.Label>
                            <Form.Control
                                value={data.genre}
                                onChange={(e) => setData({...data, genre: e.target.value})}
                                name="genre"
                                required
                                placeholder="Sci-Fi"
                                type="text"
                            />
                            <Form.Control.Feedback type="invalid">
                                Please provide genre
                            </Form.Control.Feedback>
                        </Form.Group>
                        <Form.Group>
                            <Form.Label htmlFor="relY">Year of release</Form.Label>
                            <DatePicker
                                id="relY"
                                className="form-control"
                                placeholderText="2014"
                                dateFormat="yyyy"
                                selected={data.releaseYear}
                                onChange={(e) => setData({...data, releaseYear: e})}
                                showYearPicker
                                required
                                minDate={(new Date()).setFullYear(1900)}
                            />
                            <Form.Control.Feedback type="invalid">
                                Please provide a release year
                            </Form.Control.Feedback>
                        </Form.Group>
                        <Form.Group>
                            <Form.Label>Description</Form.Label>
                            <Form.Control
                                value={data.description}
                                onChange={(e) => setData({...data, description: e.target.value})}
                                name="description"
                                required
                                placeholder="Very cool movie!"
                                as="textarea"
                                rows={3}
                            />
                            <Form.Control.Feedback type="invalid">
                                Please provide a short description
                            </Form.Control.Feedback>
                        </Form.Group>
                        <Form.Group>
                            <Form.Label>Movie poster</Form.Label>
                            <Form.File
                                className="text-nowrap text-truncate"
                                ref={fileInput}
                                custom
                                onChange={onPhotoChange}
                                label={photoLabel || method === "POST" ? "Choose photo..." : "Previous photo"}
                                accept="image/*"
                                data-browse="Add poster"
                            />
                        </Form.Group>
                        <Button
                            type="submit"
                            variant="success"
                            disabled={disabled}
                        >
                            {disabled ? "Loading..." : method === "POST" ? "Add movie" : "Update movie"}
                        </Button>
                        {photoError && <Alert variant="danger">Photo could not be uploaded. Please try again.</Alert>}
                        {resCorrect && <Alert variant="success">Movie added successfully.</Alert>}
                        {error && <Alert variant="danger">Something went wrong while adding your movie.</Alert>}
                    </Form>
                )
            }
        </Page>
    )
};

export default MovieForm;