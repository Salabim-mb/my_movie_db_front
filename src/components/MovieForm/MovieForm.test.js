import React from "react";
import MovieForm from "./MovieForm";
import {render, fireEvent, waitForElement} from '@testing-library/react';
import {paths} from "../../constants/routes";
import {createMemoryHistory} from 'history';
import {MemoryRouter, Router} from "react-router";

jest.mock("react-router-dom", () => ({
    ...jest.requireActual("react-router-dom"),
    useParams: () => ({
        id: 1
    })
}));

let reactRouterDom = require("react-router-dom");

describe("MovieForm POST", () => {
    let postFail;
    let movieData;

    beforeAll(() => {
        global.fetch = jest.fn().mockImplementation((input, init) => {
            return new Promise((resolve, reject) => {
                if (postFail) {
                    resolve({status: 500});
                } else {
                    resolve({status: 201, json: () => Promise.resolve({})});
                }
            });
        });
    });

    beforeEach(() => {
        jest.clearAllMocks();
        reactRouterDom.useParams = () => ({

        });
        movieData = {
            id: 1,
            movieName: "Interstellar",
            movieGenre: "Sci-Fi",
            movieDirector: "Christopher Nolan",
            movieReleaseDate: new Date(),
            movieImage: "",
            movieDescription: "Very cool movie!"
        }
    });

    it("should match snapshot", async () => {
        const {container, getByText} = render(
            <MemoryRouter>
                <MovieForm />
            </MemoryRouter>
        );

        await waitForElement(() => getByText("Genre"));

        expect(container).toMatchSnapshot();
    });

    it("should post movie and get response", async () => {
        const {getByLabelText, getByPlaceholderText, getByText} = render(
            <MemoryRouter>
                <MovieForm />
            </MemoryRouter>
        );

        fireEvent.change(getByLabelText("Movie title"), {
            target: {value: movieData.movieName}
        });
        fireEvent.change(getByLabelText("Director"), {
            target: {value: movieData.movieDirector}
        });
        fireEvent.change(getByLabelText("Genre"), {
            target: {value: movieData.movieGenre}
        });
        fireEvent.change(getByPlaceholderText("2014"), {
            target: { value: movieData.movieReleaseDate}
        });
        fireEvent.change(getByLabelText("Description"), {
            target: {value: movieData.movieDescription}
        });

        fireEvent.click(getByText("Add movie"));

        await waitForElement(() => getByText("Movie added successfully.", {exact: false}));

        expect(getByText("Movie added successfully", {exact: false})).toBeInTheDocument();
    });

    it("should render error on api fail", async () => {
        postFail = true;
        const {getByLabelText, getByPlaceholderText, getByText} = render(
            <MemoryRouter>
                <MovieForm />
            </MemoryRouter>
        );

        fireEvent.change(getByLabelText("Movie title"), {
            target: {value: movieData.movieName}
        });
        fireEvent.change(getByLabelText("Director"), {
            target: {value: movieData.movieDirector}
        });
        fireEvent.change(getByLabelText("Genre"), {
            target: {value: movieData.movieGenre}
        });
        fireEvent.change(getByPlaceholderText("2014"), {
            target: { value: movieData.movieReleaseDate}
        });
        fireEvent.change(getByLabelText("Description"), {
            target: {value: movieData.movieDescription}
        });

        fireEvent.click(getByText("Add movie"));

        await waitForElement(() => getByText("Something went wrong while adding your movie."));

        expect(getByText("Something went wrong while adding your movie.")).toBeInTheDocument();
    });
});

describe("MovieForm GET & PUT", () => {
    let getFail;
    let putFail;
    let movieData;



    beforeAll(() => {

        global.fetch = jest.fn().mockImplementation((input, init) => {
            return new Promise((resolve, reject) => {
                switch(init.method) {
                    case "PUT":
                        if (putFail) {
                            resolve({status: 500});
                        } else {
                            resolve({status: 200, json: () => Promise.resolve({})});
                        }
                        break;
                    case "GET":
                        if (getFail) {
                            resolve({status: 500});
                        } else {
                            resolve({status: 200, json: () => Promise.resolve(movieData)});
                        }
                        break;
                    default:
                        reject({});
                }
            });
        });
    });

    beforeEach(() => {
        jest.clearAllMocks();
        reactRouterDom.useParams = () => ({
            id: 1
        });
        movieData = {
            id: 1,
            movieName: "Interstellar",
            movieGenre: "Sci-Fi",
            movieDirector: "Christopher Nolan",
            movieReleaseDate: new Date(),
            movieImage: "",
            movieDescription: "Very cool movie!"
        }
    });

    it("should match snapshot filled", async () => {
        const {container, getByLabelText} = render(
            <MemoryRouter>
                <MovieForm />
            </MemoryRouter>
        );

        await waitForElement(() => getByLabelText("Movie title"));

        expect(container).toMatchSnapshot();
    });

    it("should render correct alert message when submitted", async () => {
        const {getByLabelText, getByText} = render(
            <MemoryRouter>
                <MovieForm />
            </MemoryRouter>
        );

        await waitForElement(() => getByLabelText("Movie title"));
        expect(getByLabelText("Movie title").value).toBe(movieData.movieName);

        fireEvent.change(getByLabelText("Movie title"), {
            target: {value: "asdf"}
        });
        fireEvent.click(getByText("Update movie"));

        await waitForElement(() => getByText("Movie updated successfully.", {exact: false}));
        expect(getByText("Movie updated successfully.", {exact: false})).toBeInTheDocument();
    });
});