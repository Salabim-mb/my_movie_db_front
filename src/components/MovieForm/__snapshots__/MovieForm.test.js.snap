// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieForm GET & PUT should match snapshot filled 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="card"
    >
      <div
        class="card-body"
      >
        <form
          class=""
          novalidate=""
        >
          <div
            class="form-group"
          >
            <label
              class="form-label"
              for="name"
            >
              Movie title
            </label>
            <input
              class="form-control"
              id="name"
              name="name"
              placeholder="Interstellar"
              required=""
              type="text"
              value="Interstellar"
            />
            <div
              class="invalid-feedback"
            >
              Please provide movie title
            </div>
          </div>
          <div
            class="form-group"
          >
            <label
              class="form-label"
              for="director"
            >
              Director
            </label>
            <input
              class="form-control"
              id="director"
              name="director"
              placeholder="Christopher Nolan"
              required=""
              type="text"
              value="Christopher Nolan"
            />
            <div
              class="invalid-feedback"
            >
              Please provide movie director
            </div>
          </div>
          <div
            class="form-group"
          >
            <label
              class="form-label"
              for="genre"
            >
              Genre
            </label>
            <input
              class="form-control"
              id="genre"
              name="genre"
              placeholder="Sci-Fi"
              required=""
              type="text"
              value="Sci-Fi"
            />
            <div
              class="invalid-feedback"
            >
              Please provide genre
            </div>
          </div>
          <div
            class="form-group"
          >
            <label
              class="form-label"
              for="releaseDate"
            >
              Year of release
            </label>
            <div
              class="react-datepicker-wrapper"
            >
              <div
                class="react-datepicker__input-container"
              >
                <input
                  class="form-control"
                  placeholder="2014"
                  required=""
                  type="text"
                  value="2020"
                />
              </div>
            </div>
            <div
              class="invalid-feedback"
            >
              Please provide a release year
            </div>
          </div>
          <div
            class="form-group"
          >
            <label
              class="form-label"
              for="description"
            >
              Description
            </label>
            <textarea
              class="form-control"
              id="description"
              name="description"
              placeholder="Very cool movie!"
              required=""
              rows="3"
            >
              Very cool movie!
            </textarea>
            <div
              class="invalid-feedback"
            >
              Please provide a short description
            </div>
          </div>
          <div
            class="form-group"
          >
            <label
              class="form-label"
              for="image"
            >
              Movie poster
            </label>
            <div
              class="text-nowrap text-truncate custom custom-file"
            >
              <input
                accept="image/*"
                class="custom-file-input"
                id="image"
                type="file"
              />
              <label
                class="custom-file-label"
                data-browse="Add poster"
                for="image"
              >
                Previous photo
              </label>
            </div>
          </div>
          <button
            class="btn btn-success"
            type="submit"
          >
            Update movie
          </button>
        </form>
      </div>
    </div>
  </div>
</div>
`;

exports[`MovieForm POST should match snapshot 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="card"
    >
      <div
        class="card-body"
      >
        <form
          class=""
          novalidate=""
        >
          <div
            class="form-group"
          >
            <label
              class="form-label"
              for="name"
            >
              Movie title
            </label>
            <input
              class="form-control"
              id="name"
              name="name"
              placeholder="Interstellar"
              required=""
              type="text"
              value=""
            />
            <div
              class="invalid-feedback"
            >
              Please provide movie title
            </div>
          </div>
          <div
            class="form-group"
          >
            <label
              class="form-label"
              for="director"
            >
              Director
            </label>
            <input
              class="form-control"
              id="director"
              name="director"
              placeholder="Christopher Nolan"
              required=""
              type="text"
              value=""
            />
            <div
              class="invalid-feedback"
            >
              Please provide movie director
            </div>
          </div>
          <div
            class="form-group"
          >
            <label
              class="form-label"
              for="genre"
            >
              Genre
            </label>
            <input
              class="form-control"
              id="genre"
              name="genre"
              placeholder="Sci-Fi"
              required=""
              type="text"
              value=""
            />
            <div
              class="invalid-feedback"
            >
              Please provide genre
            </div>
          </div>
          <div
            class="form-group"
          >
            <label
              class="form-label"
              for="releaseDate"
            >
              Year of release
            </label>
            <div
              class="react-datepicker-wrapper"
            >
              <div
                class="react-datepicker__input-container"
              >
                <input
                  class="form-control"
                  placeholder="2014"
                  required=""
                  type="text"
                  value=""
                />
              </div>
            </div>
            <div
              class="invalid-feedback"
            >
              Please provide a release year
            </div>
          </div>
          <div
            class="form-group"
          >
            <label
              class="form-label"
              for="description"
            >
              Description
            </label>
            <textarea
              class="form-control"
              id="description"
              name="description"
              placeholder="Very cool movie!"
              required=""
              rows="3"
            />
            <div
              class="invalid-feedback"
            >
              Please provide a short description
            </div>
          </div>
          <div
            class="form-group"
          >
            <label
              class="form-label"
              for="image"
            >
              Movie poster
            </label>
            <div
              class="text-nowrap text-truncate custom custom-file"
            >
              <input
                accept="image/*"
                class="custom-file-input"
                id="image"
                type="file"
              />
              <label
                class="custom-file-label"
                data-browse="Add poster"
                for="image"
              >
                Choose photo...
              </label>
            </div>
          </div>
          <button
            class="btn btn-success"
            type="submit"
          >
            Add movie
          </button>
        </form>
      </div>
    </div>
  </div>
</div>
`;
